### PR TITLE
Remove SpecOptionPartialEq

### PR DESCRIPTION
**Pull request by clubby789**
With the recent LLVM bump, the specialization for Option::partial_eq on types with niches is no longer necessary. I kept the manual implementation as it still gives us better codegen than the derive (will look at this seperately).

Also implemented PartialOrd/Ord by hand as it _somewhat_ improves codegen for #49892: https://godbolt.org/z/vx5Y6oW4Y

### CODE REVIEW

**Comment by compiler-errors** [comment]
@bors try @rust-timer queue

(are there codegen tests for this?)

**Comment by clubby789** [comment]
`tests/codegen/option-niche-eq.rs` tests this (or at least the motivating case for types with niches). I also moved the cmp::Ordering assembly test into codegen as the assumes are now eliminated and the IR comes out neatly

**Comment by clubby789** [comment]
Since I made the PR, this seems to have stopped working 😕 Rebasing onto master makes this fail locally - I think there's been a change in the pre-optimized IR emitted.
@bors r-

**Comment by clubby789** [comment]
@rustbot ready
Spelling out the match directly without a wildcard makes the opt work again in LLVM 18

File: tests/codegen/option-niche-eq.rs [hunk 1]
```diff
@@ -36,3 +34,42 @@ pub fn non_null_eq(l: Option<NonNull<u8>>, r: Option<NonNull<u8>>) -> bool {
...
+// CHECK-lABEL: @niche_eq
+#[no_mangle]
+pub fn niche_eq(l: Option<EnumWithNiche>, r: Option<EnumWithNiche>) -> bool {
+    // CHECK: start:
+    // CHECK-NEXT: icmp eq i8
+    // CHECK-NEXT: ret i1
+    l == r
+}
+
+// FIXME: This should work too
```
Comment (clubby789) [inline tests/codegen/option-niche-eq.rs:67]: Opened an issue for the poor codegen here:
https://github.com/rust-lang/rust/issues/122734
Is it okay to leave this fixme test here?

Comment (jhpratt) [reply tests/codegen/option-niche-eq.rs:67]: Yeah, that's fine.

**Pull request merged**